### PR TITLE
Implement multiple selection feature for editing tags and delete #286

### LINKED ISSUE #286: Implement multiple selection feature for editing tags and delete
#229

### PR DESCRIPTION
**Pull request by yl-coder**
Fixes #286
Fxies #291

### CODE REVIEW

File: src/main/java/address/keybindings/Bindings.java [hunk 1]
```diff
@@ -62,6 +63,7 @@ private void init(){
         FILE_NEW_ACCELERATOR = setAccelerator("FILE_NEW_ACCELERATOR", "SHORTCUT + N");
         FILE_OPEN_ACCELERATOR = setAccelerator("FILE_OPEN_ACCELERATOR", "SHORTCUT + O");
         FILE_SAVE_ACCELERATOR = setAccelerator("FILE_SAVE_ACCELERATOR", "SHORTCUT + S");
+        PERSON_TAG_ACCELERATOR = setAccelerator("PERSON_TAG_ACCELERATOR", "T");
```
Comment (damithc) [inline src/main/java/address/keybindings/Bindings.java]: `T` key binding is in between 2 `S` key bindings. :-p

Comment (yl-coder) [inline src/main/java/address/keybindings/Bindings.java]: Right. :P

**Comment by yl-coder** [comment]
Feel free to review and give suggestion now. I'm still going to add in some improvement.

On the other hand, what should the browser display when multiple card is selected? currently it will show the last selected cell(i.e. Clicking a non-selected to become selected, clicking selected to selected won't change the browser content).

List of improvement to be added:
1) Click on empty listview cell(if any) to clear the listview selection.

I will also implement multiple drag and drop. But the implementation will be abit different and complex, so i will do it in another pr.

File: src/main/java/address/controller/MainController.java [hunk 2]
```diff
@@ -237,6 +241,39 @@ private FXMLLoader loadFxml(String fxmlResourcePath) {
...
+            if (controller.isOkClicked()) {
+                return Optional.of(controller.getFinalAssignedTags());
+            }
+
+        } catch (IOException e) {
+            logger.warn("Error launching tag selection dialog: {}", e);
+            assert false : "Error loading fxml : " + FXML_TAG_SELECTION_EDIT_DIALOG;
+        }
+        return Optional.empty();
+    }
```
Comment (damithc) [inline src/main/java/address/controller/MainController.java]: Any reason to use Optional instead of an empty list?

Comment (yl-coder) [inline src/main/java/address/controller/MainController.java]: this method is to get the final tags to be updated. if an empty list is returned. the person tags will be empty. where an empty optional indicates no changes to be made

**Comment by yl-coder** [comment]
@YijinL Have a look at the changes of the Model component.

This PR is ready for review.
Will fix the drag and drop bug in another PR.

**Comment by yl-coder** [comment]
Ready for review.

File: src/main/java/address/controller/PersonOverviewController.java [hunk 3]
```diff
@@ -40,13 +42,32 @@
...
+
+    @FXML
     private ListView<ReadOnlyViewablePerson> personListView;
     @FXML
     private TextField filterField;
 
     private MainController mainController;
     private ModelManager modelManager;
 
+    private BooleanProperty isEditDisabled = new SimpleBooleanProperty(false);
```
Comment (YijinL) [inline src/main/java/address/controller/PersonOverviewController.java]: leave comment explaining this?

File: src/main/java/address/controller/PersonOverviewController.java [hunk 4]
```diff
@@ -108,6 +135,22 @@ private void handleNewPerson() {
         } while (prevInputData.isPresent() && !isAddSuccessful(prevInputData.get()));
     }
 
+    /**
+     * Called when the context menu edit is clicked.
+     */
+    private void handleEditTagPersons() {
```
Comment (YijinL) [inline src/main/java/address/controller/PersonOverviewController.java]: handleRetagPersons? no such thing as a tag-person

File: src/main/java/address/controller/PersonOverviewController.java [hunk 5]
```diff
@@ -140,16 +183,10 @@ private void handleEditPerson() {
...
-            return true;
-        } catch (DuplicatePersonException e) {
-            mainController.showAlertDialogAndWait(AlertType.WARNING, "Warning", "Cannot have duplicate person",
-                    e.toString());
-            return false;
-        }
+        modelManager.updatePerson(oldPerson, newPerson);
+        mainController.getStatusBarHeaderController().postStatus(
+                new PersonEditedStatus(new Person(oldPerson), newPerson));
+        return true;
```
Comment (YijinL) [inline src/main/java/address/controller/PersonOverviewController.java]: if always returns true, inline the method or change the method name to something that tells what it does

File: src/main/java/address/model/datatypes/person/ReadOnlyPerson.java [hunk 6]
```diff
@@ -180,4 +180,15 @@ default ReadOnlyStringProperty cityProperty() {
                 getObservableTagList()
         };
     }
+
+    static List<Tag> getCommonTags(List<ReadOnlyViewablePerson> persons) {
```
Comment (YijinL) [inline src/main/java/address/model/datatypes/person/ReadOnlyPerson.java]: `...(Collection<ReadOnlyPerson> ...`

**Comment by YijinL** [comment]
no more comments

File: src/main/java/address/controller/MainController.java (hunk 2 above)
Comment (damithc) [inline src/main/java/address/controller/MainController.java]: Got it.

Comment (damithc) [inline src/main/java/address/controller/MainController.java]: Does the whole block needs to be inside `try`?

**Comment by damithc** [comment]
Added one more comment

File: src/main/java/address/controller/MainController.java (hunk 2 above)
Comment (yl-coder) [inline src/main/java/address/controller/MainController.java]: No, only the load method. I have made the changes.

File: src/main/java/address/controller/PersonOverviewController.java (hunk 3 above)
Comment (yl-coder) [inline src/main/java/address/controller/PersonOverviewController.java]: ok noted.

File: src/main/java/address/controller/PersonOverviewController.java (hunk 4 above)
Comment (yl-coder) [inline src/main/java/address/controller/PersonOverviewController.java]: ok :p

File: src/main/java/address/controller/PersonOverviewController.java (hunk 5 above)
Comment (yl-coder) [inline src/main/java/address/controller/PersonOverviewController.java]: ok noted. changed to updatePerson()

File: src/test/java/address/unittests/keybindings/KeyBindingsManagerApiTest.java [hunk 7]
```diff
@@ -75,8 +75,6 @@ public void keyEventDetected_definedBinding_matchingEventRaised() throws Excepti
 
     @Test
     public void keyEventDetected_undefinedBinding_ignored(){
-        verifyIgnored("A");
-        verifyIgnored("G", "A");
```
Comment (yl-coder) [inline src/test/java/address/unittests/keybindings/KeyBindingsManagerApiTest.java]: @damithc Could you review this part of the changes that i just make?

Comment (yl-coder) [inline src/test/java/address/unittests/keybindings/KeyBindingsManagerApiTest.java]: I'm going to merge this first so i can continue to do the next stage of multiple select. Let me know if there is any issue, and i will fix it in the next PR.

**Pull request merged**